x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotDiskThresholdIntegTests.java

```diff
@@ -0,0 +1,287 @@
+        return false;
+    }
+
+    public void testHighWatermarkCanBeExceededOnColdOrFrozenNode() throws Exception {
+        internalCluster().startMasterOnlyNode();
+        final String dataHotNode = internalCluster().startNode(
+            Settings.builder()
+                .putList(NodeRoleSettings.NODE_ROLES_SETTING.getKey(), DiscoveryNodeRole.DATA_HOT_NODE_ROLE.roleName())
+                .build()
+        );
+
+        final var masterInfoService = (InternalClusterInfoService) internalCluster().getCurrentMasterNodeInstance(ClusterInfoService.class);
+        ClusterInfoServiceUtils.refresh(masterInfoService);
+
+        final int nbIndices = randomIntBetween(1, 5);
+        final Thread[] threads = new Thread[nbIndices];
+        final CountDownLatch latch = new CountDownLatch(nbIndices);
+
+        for (int i = 0; i < nbIndices; i++) {
+            final String index = "index-" + i;
+            var thread = new Thread(() -> {
+                try {
+                    createIndex(
+                        index,
+                        Settings.builder()
+                            .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                            .put(DataTier.TIER_PREFERENCE, DataTier.DATA_HOT)
+                            .put(INDEX_SOFT_DELETES_SETTING.getKey(), true)
+                            .put(INDEX_STORE_STATS_REFRESH_INTERVAL_SETTING.getKey(), "0ms")
+                            .put(DataTier.TIER_PREFERENCE_SETTING.getKey(), DataTier.DATA_HOT)
+                            .build()
+                    );
+                    int nbDocs = 100;
+                    try (BackgroundIndexer indexer = new BackgroundIndexer(index, client(), nbDocs)) {
+                        while (true) {
+                            waitForDocs(nbDocs, indexer);
+                            indexer.assertNoFailures();
+                            assertNoFailures(
+                                client().admin().indices().prepareForceMerge().setFlush(true).setIndices(index).setMaxNumSegments(1).get()
+                            );
+                            Map<String, Long> storeSize = sizeOfShardsStores(index);
+                            if (storeSize.get(index) > WATERMARK_BYTES) {
+                                break;
+                            }
+                            int moreDocs = scaledRandomIntBetween(100, 1_000);
+                            indexer.continueIndexing(moreDocs);
+                            nbDocs += moreDocs;
+                        }
+                    } catch (Exception e) {
+                        throw new AssertionError(e);
+                    }
+                } finally {
+                    latch.countDown();
+                }
+            });
+            threads[i] = thread;
+            thread.start();
+        }
+
+        latch.await();
+        for (Thread thread : threads) {
```

Comment:
I think joining the threads is unnecessary here. The latch already makes sure that all the threads have finished.

Reply:
Also in that case we don't need to collect the threads to an array.

Reply:
Ievgen made the same comment :) Ok you both won on this, I'll remove the join